### ISSUE TITLE
[Visual impairment accessibility] meta-issue

### ISSUE THREAD

**Issue by joshuabengal** [Cockatrice/Cockatrice#2388]
<b>OS:</b>
Window 7

<b>Cockatrice version:</b>
Custom

___
<br>

I have a custom build of Cockatrice I built due to vision related disabilities and eye strain caused by color choices made in the official build. I tried to log in today and get a connection error saying I needed to upgrade my client to connect. 

Instead of getting the latest code and having to go thorough all the various UI changes I need to make in order to use the program, can anyone point out the changes that I could make/files I could replace that would enable my client to connect again? 

Thanks

**Comment by skwerlman** [comment]
really, we should just start using qss to style the program...

**Comment by Daenyth** [comment]
cc @woogerboy21 what is being enforced now?

@joshuabengal I'd really like to get the official client in a state where you can use it without modification. Can you elaborate on some of the specific problems? It may be that we can actually solve them via use of the built-in theme system - it allows using CSS to style much of the application.

**Comment by woogerboy21** [comment]
The woogerworks server started to enforce the idle_client feature due to a correction in one of the event protocol components.  On older clients when the idle_client warning is sent they would receive a message saying they were promoted to "administrator" rather than the warning dialog for "your about to be disconnected due to being idle".

Its a pretty important enforcement.  Otherwise I wouldn't have required it.

**Comment by joshuabengal** [comment]
> The woogerworks server started to enforce the idle_client feature due to a correction in one of the event protocol components. On older clients when the idle_client warning is sent they would receive a message saying they were promoted to "administrator" rather than the warning dialog for "your about to be disconnected due to being idle".

Understood. Can you let me know what files/changes need to be made in the client to accommodate this requirement and let it connect? Most of my familiarity with the codebase is in on the presentation side. 

> @joshuabengal I'd really like to get the official client in a state where you can use it without modification. Can you elaborate on some of the specific problems?

Unfortunately I doubt that would be feasible. Visual triggers are extremely individual, what triggers me won't necessarily trigger anyone else with a similar disability. As for themeing the application in CSS, you would need to apply a class to every single element in the entire app. Border styles, title bar widths, anti aliasing filters, even undefined areas inside other windows. It would likely be far to much effort. 

For example, there was an update sometime in the past, I think it was poixen who made he commit if I was not mistaken, that darkened the interior box inside the zoneviewzone widget. Previously when you viewed a zone, the background of that window was all the same color gray, now it was a dark grey box inside a lighter gray box. When I view different greys next to each other of a certain RGB value I get a severe eye strain trigger, so this change rendered the client unusable. The same things happen when changing border elements with different anti aliasing (due to the dithering performed by the graphics card to display the intermediate colors). That change was also just made to the program with no warning, rendering it unusable for me for a time. In the same vein removing the larger "Close" button and changing it to a much smaller X triggers RSI induced wrist and thumb pain due to the finer tracking required to hit the button.

Given the case that devs were free to modify visual elements pretty much on a whim and publish them to production, I decided it would be much easier for me to grab a stable version of the code and go through removing the various visual effects that were causing me pain. Altogether I believe it was something like 15 or so changes that need to be made. And it worked great, until the server started rejecting my log in, so I'd like to update my custom client to be able to comply with the new server requirement. It would just be far easier for me to make a custom client that work vs trying to change the main codebase into one that doesn't trigger me. 

I do think it would be a good idea for the other devs to abstain from making visual changes to the client and publishing them without any notice. It's a big problem for folks with visual perception disabilities like myself ,and other sufferers may not have the coding skills to implement my workarounds. Just something to think about.

**Comment by woogerboy21** [comment]
Make the common/pb/event_notify_user.proto look like the one on the most recent development site and then make the common/featureset.cpp look like the one on the most recent development site and recompile.  That should do it.

Otherwise if the protocol file is out of date you will see a message about being upgraded to an "admin" when the idle client warning is sent.

You really should consider figuring out how to handle situations like this as I cant say in the future what other updates may be enforced.

**Comment by joshuabengal** [comment]
> Make the common/pb/event_notify_user.proto look like the one on the most recent development site and then make the common/featureset.cpp look like the one on the most recent development site and recompile. That should do it.

Thanks, I will give it a shot

> You really should consider figuring out how to handle situations like this as I cant say in the future what other updates may be enforced.

Trust me, a cure for my disability is first and foremost on my mind at all times. Until that day comes, I cannot "turn off" my sensitivity, so workarounds are the only way forward. 
\> As for the points to the visual aspects. If you choose to embark on your own mass modification / customization its not the fault of this group to support your modifications. Don't take that in a wrong way.

**Comment by Daenyth** [comment]
@woogerboy21 I think you're being a bit overzealous. It's fine if he modifies his client; everything he's suggested has been 100% reasonable.

@joshuabengal it's a good point about different people needing different things, sight-wise. That being said, we can only go off what's reported, and you're one of only a few reporting accessibility issues, so I'd still weight your suggestions pretty highly. I actually was already wanting to get to a point where everything in the client has css classes, so what you said isn't really a barrier, just work that we wanted to do anyway.

Ultimately I think we will continue making UI changes, but accessibility impact is something we can start considering in the process. If you could categorize classes of changes that are problematic, at least for you, that might help us. Personally I have no experience with software accessibility, and I'm not sure anyone else contributing does either.

Is your fork of trice on here? If you publish your patches we can probably look at them and potentially find a way to start incorporating some of them, maybe as options.

cc @poixen some good ideas here to consider.

**Comment by woogerboy21** [comment]
@Daenyth overzealous? I can easily see a scenario were some one goes down this path and something as simple as a new zone is created resulting in a new branch of logic that would render the "custom" client useless and applying the entire set of logic branch(s) would become tedious.  Yet the user requests that we don't allow UI changes because there client has been customized so drastically they cant simply upgrade it?

I think you simply miss-read my intentions with the comment.  I was merely trying to suggest that if the user has produced such a change to the client it would beneficial for them (and others) to create a PR allowing for those changes to be rolled in resulting in future support-ability.  At the very least if they didnt submit the change as a PR they could have a local copy of the source that they could update their master branch with the projects master and see were the conflicts (if any) occur.

**Comment by joshuabengal** [comment]
>  That being said, we can only go off what's reported, and you're one of only a few reporting accessibility issues, so I'd still weight your suggestions pretty highly. I actually was already wanting to get to a point where everything in the client has css classes, so what you said isn't really a barrier, just work that we wanted to do anyway.

Glad to share, although as I said before these things are highly individualized, so these are just what I have found effects me. 

The majority of my changes are color related, especially around grays. For people sensitive to eye strain caused by colors, putting colors that are close to each other on the RGB scale on the same area of the screen can cause eye strain. In my cause changing the single gray background to a dual gray background made the program unusable. I also saw in later versions there seemed to be wholesale replacements of the default gray to a lighter shade of gray. Colors in the range of RGB 245, 245, 245 - 253, 253, 253 are notorious for being a very strong eye strain and facial pain trigger for sensitive individuals. For that reason I used as my starting point a build which had the standard Windows gray colored controls. Nearly all my changes are color related, usually replacing grays with deeper grays. 

Also, there are several hard coded colors throughout the program, this interferes with Windows themes, including the High Contrast theme designed to help those with visual eye strain triggers. 

Also, changing large hitbox items to small hitbox items is a challenge to folks suffering from RSI's. I posted about this before as an issue, the removal of the "Close" button from the zoneviewzone widget and instead having a small 10px "X" button as the close button greatly increased the tracking precision needed to close that window. I had to add that Close button back in to make the window controls comfortable to use. 

I do not have a fork on Github, I downloaded the source code of an earlier versions and have been making all my changes locally, I'm afraid I'm not much of a Github expert. 

> Yet the user requests that we don't allow UI changes because there client has been customized so drastically they cant simply upgrade it?

**Comment by Daenyth** [comment]
Thanks, that's really useful information.

> there are several hard coded colors throughout the program

I already wanted to get rid of those; if you can provide a list of locations in a new ticket, it would help us clean them up faster.

>  changing large hitbox items to small hitbox items is a challenge 

That's good feedback, we can incorporate that.

> the removal of the "Close" button from the zoneviewzone widget and instead having a small 10px "X" button as the close button greatly increased the tracking precision needed to close that window

I remember that vaguely; AFAIK the esc key also works to close the view - is that not the case? Or if so, is that still a strain issue?

I wonder if it's possible to add some configuration that would control button sizing. Maybe we could have it default to the current type but have the size be variable on configuration.

I have more thoughts on this stuff, but not time at the moment to put them all down - please continue to leave feedback for us, it's a huge boon to the project to have an invested user like yourself. It's challenging for us to consider these issues, since AFAIK none of the common contributors suffer from issues like this.

> I do not have a fork on Github, I downloaded the source code of an earlier versions and have been making all my changes locally, I'm afraid I'm not much of a Github expert.

I'd be more than happy to help walk you through getting set up - it's pretty easy. PM me on our [gitter chat](https://gitter.im/Cockatrice/Cockatrice) and I'll reply when I'm available, if you're interested.

**Comment by Daenyth** [comment]
There's no clear next step here but I want to leave this open for now so we don't entirely forget about it.

**Comment by joshuabengal** [comment]
Just as a point of follow up I followed Woogerboys recommendation of grabbing those 2 files and recompiling, and every is once again working, so thanks for that!

**Comment by Daenyth** [comment]
Useful https://tallys.github.io/color-theory/

**Comment by ZeldaZach** [comment]
Hey @joshuabengal if you ever want to come back and help improve the accessibility of Cockatrice, I'd love to get your opinion. However, it's been 3 years since last comment and I'll be closing in favor of #4133
